File: app/src/lib/gateway-session.ts
type CachedToken = { value: string; proof: string; expiresAt: number };

let cachedToken: CachedToken | null = null;
let pending: Promise<CachedToken> | null = null;

const TOKEN_STORAGE_KEY = "gateway.session.token";

const storage = (() => {
  try {
    const store = window.localStorage;
    store.getItem(TOKEN_STORAGE_KEY);
    return store;
  } catch {
    return null;
  }
})();

function readToken(): CachedToken | null {
  if (!storage) return null;
  try {
    const raw = storage.getItem(TOKEN_STORAGE_KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    if (
      parsed &&
      typeof parsed.value === "string" &&
      typeof parsed.proof === "string" &&
      typeof parsed.expiresAt === "number"
    ) {
      return parsed;
    }
  } catch {
    // ignore
  }
  return null;
}

function persist(token: CachedToken | null) {
  if (!storage) return;
  try {
    if (token) {
      storage.setItem(TOKEN_STORAGE_KEY, JSON.stringify(token));
    } else {
      storage.removeItem(TOKEN_STORAGE_KEY);
    }
  } catch {
    // ignore
  }
}

const stored = readToken();
if (stored && stored.expiresAt > Date.now()) {
  cachedToken = stored;
}

const SESSION_ENDPOINT = "/api/session";

async function requestSession(): Promise<CachedToken> {
  const resp = await fetch(SESSION_ENDPOINT, {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
  });
  if (!resp.ok) throw new Error(`session status ${resp.status}`);
  const payload = (await resp.json()) as { csrfToken?: string; csrfProof?: string; expiresAt?: number };
  if (!payload.csrfToken || !payload.csrfProof) throw new Error("missing csrf");
  const token = {
    value: payload.csrfToken,
    proof: payload.csrfProof,
    expiresAt: payload.expiresAt ?? Date.now() + 1000 * 60 * 20,
  };
  cachedToken = token;
  persist(token);
  return token;
}

export async function ensureGatewaySession(): Promise<{ token: string; proof: string }> {
  if (cachedToken && cachedToken.expiresAt - Date.now() > 30_000) {
    return { token: cachedToken.value, proof: cachedToken.proof };
  }
  if (!pending) {
    pending = requestSession().finally(() => {
      pending = null;
    });
  }
  const session = await pending;
  return { token: session.value, proof: session.proof };
}

export async function authorizedFetch(input: RequestInfo | URL, init: RequestInit = {}) {
  const session = await ensureGatewaySession();
  const headers = new Headers(init.headers ?? {});
  headers.set("X-Gateway-CSRF", session.token);
  headers.set("X-Gateway-CSRF-Proof", session.proof);
  const finalInit: RequestInit = { ...init, headers, credentials: "include" };
  const resp = await fetch(input, finalInit);
  if (resp.status === 401 || resp.status === 403) {
    cachedToken = null;
    persist(null);
    const fresh = await ensureGatewaySession();
    headers.set("X-Gateway-CSRF", fresh.token);
    headers.set("X-Gateway-CSRF-Proof", fresh.proof);
    return fetch(input, { ...init, headers, credentials: "include" });
  }
  return resp;
}

File: dioxus-app/src/gateway_session.rs
use gloo_net::http::Request;
use web_sys::RequestCredentials;
use gloo_storage::{LocalStorage, Storage};
use serde::{Deserialize, Serialize};
use serde::de::DeserializeOwned;

const TOKEN_STORAGE_KEY: &str = "gateway.session.token";
const SESSION_ENDPOINT: &str = "/api/session";

#[derive(Clone, Debug, Serialize, Deserialize)]
struct CachedToken {
    value: String,
    proof: String,
    expires_at: i64,
}

fn read_cached_token() -> Option<CachedToken> {
    LocalStorage::get(TOKEN_STORAGE_KEY).ok()
}

fn write_cached_token(token: Option<&CachedToken>) {
    match token {
        Some(value) => {
            let _ = LocalStorage::set(TOKEN_STORAGE_KEY, value);
        }
        None => {
            LocalStorage::delete(TOKEN_STORAGE_KEY);
        }
    }
}

fn is_valid(token: &CachedToken) -> bool {
    if token.value.is_empty() || token.proof.is_empty() {
        return false;
    }
    let now = js_sys::Date::now() as i64;
    token.expires_at - now > 30_000
}

async fn request_new_session() -> Result<CachedToken, String> {
    let response = Request::post(SESSION_ENDPOINT)
        .header("Content-Type", "application/json")
        .credentials(RequestCredentials::Include)
        .send()
        .await
        .map_err(|err| format!("session request failed: {err}"))?;

    if !response.ok() {
        return Err(format!("session request failed: {}", response.status()));
    }

    #[derive(Deserialize)]
    struct SessionPayload {
        #[serde(rename = "csrfToken")]
        csrf_token: Option<String>,
        #[serde(rename = "csrfProof")]
        csrf_proof: Option<String>,
        #[serde(rename = "expiresAt")]
        expires_at: Option<i64>,
    }

    let payload = response
        .json::<SessionPayload>()
        .await
        .map_err(|err| format!("session decode failed: {err}"))?;

    let csrf_token = payload.csrf_token.ok_or("session missing csrfToken")?;
    let csrf_proof = payload.csrf_proof.ok_or("session missing csrfProof")?;
    let expires_at = payload
        .expires_at
        .unwrap_or_else(|| (js_sys::Date::now() as i64) + 1000 * 60 * 30);

    let token = CachedToken {
        value: csrf_token,
        proof: csrf_proof,
        expires_at,
    };
    write_cached_token(Some(&token));
    Ok(token)
}

pub async fn ensure_gateway_session() -> Result<(String, String), String> {
    if let Some(token) = read_cached_token() {
        if is_valid(&token) {
            return Ok((token.value, token.proof));
        }
        write_cached_token(None);
    }

    let token = request_new_session().await?;
    Ok((token.value, token.proof))
}

pub async fn authorized_post(url: &str, body: &str) -> Result<gloo_net::http::Response, String> {
    let (token, proof) = ensure_gateway_session().await?;
    let request = Request::post(url)
        .header("Content-Type", "application/json")
        .header("X-Gateway-CSRF", &token)
        .header("X-Gateway-CSRF-Proof", &proof)
        .credentials(RequestCredentials::Include)
        .body(body)
        .map_err(|err| format!("request failed: {err}"))?;
    let response = request
        .send()
        .await
        .map_err(|err| format!("request failed: {err}"))?;

    if response.status() == 401 || response.status() == 403 {
        write_cached_token(None);
    }
    Ok(response)
}

pub async fn authorized_get_with_headers(
    url: &str,
    headers: &[(String, String)],
) -> Result<gloo_net::http::Response, String> {
    let _ = ensure_gateway_session().await;
    let mut request = Request::get(url).credentials(RequestCredentials::Include);
    for (key, value) in headers {
        request = request.header(key, value);
    }
    let response = request
        .send()
        .await
        .map_err(|err| format!("request failed: {err}"))?;

    if response.status() == 401 || response.status() == 403 {
        write_cached_token(None);
        let _ = ensure_gateway_session().await;
        let mut retry_request = Request::get(url).credentials(RequestCredentials::Include);
        for (key, value) in headers {
            retry_request = retry_request.header(key, value);
        }
        let retry = retry_request
            .send()
            .await
            .map_err(|err| format!("request failed: {err}"))?;
        if retry.status() == 401 || retry.status() == 403 {
            write_cached_token(None);
        }
        return Ok(retry);
    }
    Ok(response)
}

pub async fn authorized_get_json_with_headers<T: DeserializeOwned>(
    url: &str,
    headers: &[(String, String)],
) -> Result<T, String> {
    let response = authorized_get_with_headers(url, headers).await?;
    if !response.ok() {
        return Err(format!("http {}", response.status()));
    }
    response
        .json::<T>()
        .await
        .map_err(|err| format!("decode failed: {err}"))
}
